src/EnergyPlus/SurfaceOctree.cc

```diff
@@ -0,0 +1,265 @@
+
+		// Branch sub-tree
+		branch();
+	}
+
+	// Valid?
+	bool
+	SurfaceOctreeCube::
+	valid() const
+	{
+		if ( le( l_, c_ ) && le( c_, u_ ) ) {
+			Real const tol2( std::max( std::max( ObjexxFCL::magnitude_squared( l_ ), ObjexxFCL::magnitude_squared( u_ ) ) * ( 4 * std::numeric_limits< Real >::epsilon() ), 2 * std::numeric_limits< Real >::min() ) );
+			if ( ObjexxFCL::distance_squared( c_, cen( l_, u_ ) ) <= tol2 ) {
+				Real const tol( std::max( std::sqrt( std::max( ObjexxFCL::magnitude_squared( l_ ), ObjexxFCL::magnitude_squared( u_ ) ) ) * ( 4 * std::numeric_limits< Real >::epsilon() ), 2 * std::numeric_limits< Real >::min() ) );
+				Vertex const d( u_ - l_ ); // Diagonal
+				return ( std::abs( d.x - w_ ) <= tol ) && ( std::abs( d.x - d.y ) <= tol ) && ( std::abs( d.x - d.z ) <= tol ); // Uniform side widths?
+			}
+		}
+		return false;
+	}
+
+	// Branch to Sub-Tree
+	void
+	SurfaceOctreeCube::
+	branch()
+	{
+		if ( ( surfaces_.size() > maxSurfaces_ ) && ( d_ < maxDepth_ ) ) {
+			// Assign Surfaces to cubes containing them
+			Surfaces surfaces_all;
+			surfaces_all.swap( surfaces_ );
+			for ( auto * surface_p : surfaces_all ) { // Surfaces
+				surfaceBranch( *surface_p );
+			}
+
+			// Compact occupied cube array
+			n_ = 0u;
+			for ( std::uint8_t i = 0; i < 8; ++i ) {
+				SurfaceOctreeCube * & cube = cubes_[ i ];
+				if ( cube != nullptr ) {
+					if ( n_ < i ) {
+						cubes_[ n_ ] = cube;
+						cube = nullptr;
+					}
+					++n_;
+				}
+			}
+
+			// Branch sub-tree recursively
+			for ( std::uint8_t i = 0; i < n_; ++i ) {
+				cubes_[ i ]->branch();
+			}
+		}
+	}
+
+	// Surface Branch Processing
+	void
+	SurfaceOctreeCube::
+	surfaceBranch( Surface & surface )
+	{
+		Real const h( 0.5 * w_ ); // Half-width
+		Vertex sl( 0.0 ), su( 0.0 ); // Surface bounding box corners
+		sl = su = surface.Vertex[ 0 ]; // Initialize corners to first vertex of first surface
+		auto const & vertices( surface.Vertex ); // Surface vertices
+		for ( auto const & vertex : vertices ) { // Expand bounding box to hold surface vertices
+			sl.min( vertex );
+			su.max( vertex );
+		}
+		Vertex const ctr( cen( sl, su ) );
+		std::uint8_t const i( ctr.x <= c_.x ? 0 : 1 );
+		std::uint8_t const j( ctr.y <= c_.y ? 0 : 1 );
+		std::uint8_t const k( ctr.z <= c_.z ? 0 : 1 );
+		SurfaceOctreeCube * & cube = cubes_[ ( i << 2 ) + ( j << 1 ) + k ];
+		if ( cube != nullptr ) { // Candidate cube exists
+			if ( le( cube->l_, sl ) && le( su, cube->u_ ) ) { // Surface is contained in sub-cube
+				cube->add( surface );
+			} else { // Surface stays in this cube
+				surfaces_.push_back( &surface );
+			}
+		} else { // Create cube if surface contained
+			Real const x( i * h );
+			Real const y( j * h );
+			Real const z( k * h );
+			Vertex const l( l_.x + x, l_.y + y, l_.z + z );
+			Vertex const u( c_.x + x, c_.y + y, c_.z + z );
+			if ( le( l, sl ) && le( su, u ) ) { // Surface is contained in sub-cube
+				cube = new SurfaceOctreeCube( d_ + 1, l, u, h );
```

Comment:
@DeadParrot Like I said, you are cleaning it up and it stays within the class. I am less adverse to pointers in performance critical code than the rest of the team probably. 

I am mostly pointing this out for emphasis and posterity so there is a record of why you are using c-style array and pointer here.